rholang/src/main/scala/rholang/rosette/Equivalence.scala

```diff
@@ -0,0 +1,279 @@
+      case (lift1: PLift, lift2: PLift) => {
+        nameEquivalent(env1, lift1.chan_, env2, lift2.chan_) &&
+          allStructurallyEquivalent(env1, lift1.listproc_, env2, lift2.listproc_)
+      }
+      case (input1: PInput, input2: PInput) =>
+        allBindsEquivalent (
+          env1, input1.listbind_.asScala.toList,
+          env2, input2.listbind_.asScala.toList
+        ) match {
+          case None => false
+          case Some((names1,names2)) =>
+            allCPatternEquivalent(env1,names1,env2,names2) match {
+              case None => false
+              case Some((newEnv1,newEnv2)) =>
+                structurallyEquivalent(newEnv1, input1.proc_, newEnv2, input2.proc_)
+            }
+        }
+      case (_: PChoice, _: PChoice) => ???
+      case (_: PMatch, _: PMatch) => ???
+      case (_: PNew, _: PNew) => ???
+      case (print1: PPrint, print2: PPrint) =>
+        structurallyEquivalent(env1, print1.proc_, env2, print2.proc_)
+      case (constr1: PConstr, constr2: PConstr) => {
+        env1.equivalent(constr1.var_, env2, constr2.var_) &&
+          allStructurallyEquivalent(env1, constr1.listproc_, env2, constr2.listproc_)
+      }
+      case (contr1: PContr, contr2: PContr) => {
+        allCPatternEquivalent(
+          env1, contr1.listcpattern_.asScala.toList,
+          env2, contr2.listcpattern_.asScala.toList
+          ) match {
+            case None => false
+            case Some((newenv1,newenv2)) =>
+              env1.equivalent(contr1.var_, env2, contr2.var_) &&
+                structurallyEquivalent(newenv1, contr1.proc_, newenv2, contr2.proc_)
+          }
+      }
+      case (p1: PPar, p2: PPar) => parEquiv(parLeaves(p1),parLeaves(p2))
+      case _ => false
+    }
+  }
+
+  def bindEquivalent(env1: DeBruijn, b1: Bind, env2: DeBruijn, b2: Bind): Option[(CPattern,CPattern)] =
+    (b1, b2) match {
+      case (inpBind1: InputBind, inpBind2: InputBind) =>
+        if (nameEquivalent(env1, inpBind1.chan_, env2, inpBind2.chan_)) {
+          Some((inpBind1.cpattern_, inpBind2.cpattern_))
+        } else None
+      case (condInpBind1: CondInputBind, condInpBind2: CondInputBind) =>
+        val check =
+          nameEquivalent(env1, condInpBind1.chan_, env2, condInpBind2.chan_) &&
+          structurallyEquivalent(env1, condInpBind1.proc_, env2, condInpBind2.proc_)
+        if (check) {
+          Some((condInpBind1.cpattern_, condInpBind2.cpattern_))
+        } else None
+      case _ => None 
+    }
+
+  def allBindsEquivalent(env1: DeBruijn, b1: List[Bind], env2: DeBruijn, b2: List[Bind]): Option[(List[CPattern],List[CPattern])] =
+    b1 match {
+      case Nil => b2 match {
+        case Nil => Some((Nil,Nil))
+        case _ => None
+      }
+      case head :: tail => b2.partition {
+        bind => !(bindEquivalent(env1,bind,env2,head).isEmpty)
+      } match {
+        case (Nil,_) => None
+        case (eqhd,eqtl) =>
+          val init = (
```

Comment:
I think it is worth providing a structure here for readability & maintainability purposes. This tuple really tells my anything. I have to navigate whole fold to understand difference between `List[CPattern]` on `_2` and `_3`.